File: assessment-trainer/lib/vision/js/detect.ts
/**
 * Matrix / cell / answer-option detection — port of python/vision/detect.py.
 * Cell interiors are found as enclosed, rectangular background regions.
 */
import { label, type Gray } from "./raster";

export interface Box {
  x: number;
  y: number;
  w: number;
  h: number;
}
const cxOf = (b: Box) => b.x + b.w / 2;
const cyOf = (b: Box) => b.y + b.h / 2;
const area = (b: Box) => b.w * b.h;
const contains = (a: Box, o: Box, pad = 2) => a.x - pad <= o.x && a.y - pad <= o.y && o.x + o.w <= a.x + a.w + pad && o.y + o.h <= a.y + a.h + pad;
function iou(a: Box, o: Box) {
  const ix = Math.max(0, Math.min(a.x + a.w, o.x + o.w) - Math.max(a.x, o.x));
  const iy = Math.max(0, Math.min(a.y + a.h, o.y + o.h) - Math.max(a.y, o.y));
  const inter = ix * iy;
  return inter / (area(a) + area(o) - inter);
}
const median = (xs: number[]) => {
  const s = [...xs].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
};

export class DetectionError extends Error {
  constructor(public stage: string, message: string, public region: Box | null = null, public boxes: Box[] = []) {
    super(message);
  }
}

/** Candidate boxes: enclosed background regions that are square and rectangular. */
export function squareBoxes(ink: Uint8Array, g: Gray): Box[] {
  const { w, h } = g;
  const minSide = Math.max(20, Math.floor(0.03 * Math.min(w, h)));
  const maxSide = Math.floor(0.6 * Math.min(w, h));
  const { labels, comps } = label(ink, w, h, 0, 4);
  const out: Box[] = [];
  for (const c of comps) {
    if (c.touchesBorder) continue;
    const bw = c.maxx - c.minx + 1;
    const bh = c.maxy - c.miny + 1;
    if (bw < minSide || bh < minSide || bw > maxSide || bh > maxSide) continue;
    if (bw / bh < 0.85 || bw / bh > 1.18) continue;
    if (c.count < 0.3 * bw * bh) continue; // thin rings between double lines
    // Straight edges: the region must fill most of its bounding-box border.
    let own = 0;
    let tot = 0;
    // Check one pixel inside the box: anti-aliased border pixels are ignored.
    const [x0, x1, y0, y1] = [c.minx + 1, c.maxx - 1, c.miny + 1, c.maxy - 1];
    for (let x = x0; x <= x1; x++) {
      tot += 2;
      if (labels[y0 * w + x] === c.label) own++;
      if (labels[y1 * w + x] === c.label) own++;
    }
    for (let y = y0; y <= y1; y++) {
      tot += 2;
      if (labels[y * w + x0] === c.label) own++;
      if (labels[y * w + x1] === c.label) own++;
    }
    if (own / tot < 0.8) continue;
    out.push({ x: c.minx, y: c.miny, w: bw, h: bh });
  }
  out.sort((a, b) => area(a) - area(b));
  const dedup: Box[] = [];
  for (const b of out) if (!dedup.some((o) => iou(b, o) > 0.75)) dedup.push(b);
  return dedup;
}

function cluster1d(values: number[], tol: number): number[] {
  const groups: number[][] = [];
  for (const v of [...values].sort((a, b) => a - b)) {
    const last = groups[groups.length - 1];
    if (last && Math.abs(v - last.reduce((s, x) => s + x, 0) / last.length) <= tol) last.push(v);
    else groups.push([v]);
  }
  return groups.map((g) => g.reduce((s, x) => s + x, 0) / g.length);
}

function sizeGroups(boxes: Box[]): Box[][] {
  const groups: Box[][] = [];
  for (const b of [...boxes].sort((a, c) => c.w - a.w)) {
    const g = groups.find((gr) => {
      const rw = median(gr.map((x) => x.w));
      const rh = median(gr.map((x) => x.h));
      return Math.abs(b.w - rw) <= 0.1 * rw && Math.abs(b.h - rh) <= 0.1 * rw;
    });
    if (g) g.push(b);
    else groups.push([b]);
  }
  return groups.filter((g) => g.length >= 2);
}

export interface Lattice {
  rows: number;
  cols: number;
  xs: number[];
  ys: number[];
  size: number;
  cells: Map<string, Box>;
}

export const latticeMissing = (l: Lattice) => {
  const out: [number, number][] = [];
  for (let r = 0; r < l.rows; r++) for (let c = 0; c < l.cols; c++) if (!l.cells.has(`${r},${c}`)) out.push([r, c]);
  return out;
};
export const slotBox = (l: Lattice, r: number, c: number): Box => {
  const s = Math.round(l.size);
  return { x: Math.round(l.xs[c] - s / 2), y: Math.round(l.ys[r] - s / 2), w: s, h: s };
};
export const latticeBox = (l: Lattice): Box => {
  const s = l.size;
  const x0 = Math.min(...l.xs) - s / 2;
  const y0 = Math.min(...l.ys) - s / 2;
  return { x: Math.round(x0), y: Math.round(y0), w: Math.round(Math.max(...l.xs) - Math.min(...l.xs) + s), h: Math.round(Math.max(...l.ys) - Math.min(...l.ys) + s) };
};

function fitLattice(group: Box[]): Lattice {
  const size = median(group.map((b) => b.w));
  const tol = 0.25 * size;
  const fill = (centers: number[]) => {
    if (centers.length < 2) return centers;
    let pitch = Infinity;
    for (let i = 1; i < centers.length; i++) pitch = Math.min(pitch, centers[i] - centers[i - 1]);
    const out = [centers[0]];
    for (const v of centers.slice(1)) {
      while (v - out[out.length - 1] > 1.5 * pitch) out.push(out[out.length - 1] + pitch);
      out.push(v);
    }
    return out;
  };
  const xs = fill(cluster1d(group.map(cxOf), tol));
  const ys = fill(cluster1d(group.map(cyOf), tol));
  const cells = new Map<string, Box>();
  for (const b of group) {
    let c = 0;
    let r = 0;
    xs.forEach((x, i) => Math.abs(cxOf(b) - x) < Math.abs(cxOf(b) - xs[c]) && (c = i));
    ys.forEach((y, i) => Math.abs(cyOf(b) - y) < Math.abs(cyOf(b) - ys[r]) && (r = i));
    if (Math.abs(cxOf(b) - xs[c]) <= tol && Math.abs(cyOf(b) - ys[r]) <= tol) cells.set(`${r},${c}`, b);
  }
  return { rows: ys.length, cols: xs.length, xs, ys, size, cells };
}

function pitchRatio(l: Lattice): number {
  const p: number[] = [];
  const diffs = (a: number[]) => a.slice(1).map((v, i) => v - a[i]);
  if (l.xs.length > 1) p.push(median(diffs(l.xs)));
  if (l.ys.length > 1) p.push(median(diffs(l.ys)));
  return p.length ? Math.min(...p) / l.size : 99;
}

export interface Layout {
  matrix: Lattice;
  options: Box[];
  missing: [number, number] | null;
  inferredOptions: number;
}

export function detectLayout(ink: Uint8Array, g: Gray): Layout {
  const boxes = squareBoxes(ink, g);
  if (boxes.length < 4) throw new DetectionError("matrix", "Too few box-like regions found; no matrix grid visible.", null, boxes);
  const groups = sizeGroups(boxes);
  let best: Lattice | null = null;
  let bestGroup: Box[] | null = null;
  for (const gr of groups) {
    const lat = fitLattice(gr);
    if (lat.rows * lat.cols < 4 || lat.rows < 2 || lat.cols < 2) continue;
    if (pitchRatio(lat) > 1.35) continue;
    if (lat.cells.size < lat.rows * lat.cols - 2) continue;
    if (!best || lat.size * lat.cells.size > best.size * best.cells.size) {
      best = lat;
      bestGroup = gr;
    }
  }
  if (!best) {
    const region = groups.length ? groups.reduce((a, b) => (b.reduce((s, x) => s + area(x), 0) > a.reduce((s, x) => s + area(x), 0) ? b : a)) : boxes;
    const x0 = Math.min(...region.map((b) => b.x));
    const y0 = Math.min(...region.map((b) => b.y));
    const x1 = Math.max(...region.map((b) => b.x + b.w));
    const y1 = Math.max(...region.map((b) => b.y + b.h));
    throw new DetectionError("matrix", "No tightly packed grid of equally sized cells was found.", { x: x0, y: y0, w: x1 - x0, h: y1 - y0 }, boxes);
  }
  const mb = latticeBox(best);
  let optBest: Box[] | null = null;
  for (let gr of groups) {
    if (gr === bestGroup) continue;
    gr = gr.filter((b) => !contains(mb, b, 6));
    if (gr.length < 3 || gr.length > 8) continue;
    if (gr.some((b) => boxes.some((o) => contains(o, b, 2) && area(o) > 1.5 * area(b)))) continue;
    const rows = cluster1d(gr.map(cyOf), 0.3 * median(gr.map((b) => b.h)));
    if (rows.length > 2) continue;
    if (!optBest || median(gr.map((b) => b.w)) > median(optBest.map((b) => b.w))) optBest = gr;
  }
  if (!optBest) throw new DetectionError("options", "The matrix was found but no row of answer options was detected.", mb, boxes);
  const rows = cluster1d(optBest.map(cyOf), 0.3 * median(optBest.map((b) => b.h)));
  const rowOf = (b: Box) => rows.reduce((bi, r, i) => (Math.abs(cyOf(b) - r) < Math.abs(cyOf(b) - rows[bi]) ? i : bi), 0);
  const sorted = [...optBest].sort((a, b) => rowOf(a) - rowOf(b) || cxOf(a) - cxOf(b));
  // Fill gaps in evenly spaced option rows (a broken option border).
  const options: Box[] = [];
  let inferred = 0;
  for (let ri = 0; ri < rows.length; ri++) {
    const row = sorted.filter((b) => rowOf(b) === ri);
    if (row.length < 2) {
      options.push(...row);
      continue;
    }
    let pitch = Infinity;
    for (let i = 1; i < row.length; i++) pitch = Math.min(pitch, cxOf(row[i]) - cxOf(row[i - 1]));
    const filled = [row[0]];
    for (const b of row.slice(1)) {
      while (cxOf(b) - cxOf(filled[filled.length - 1]) > 1.5 * pitch) {
        const p = filled[filled.length - 1];
        filled.push({ x: Math.round(p.x + pitch), y: p.y, w: p.w, h: p.h });
        inferred++;
      }
      filled.push(b);
    }
    options.push(...filled);
  }
  const missing = latticeMissing(best);
  return { matrix: best, options, missing: missing.length === 1 ? missing[0] : null, inferredOptions: inferred };
}

File: assessment-trainer/python/vision/detect.py
"""Matrix, cell and answer-option detection.

Works on screenshots that also contain browser chrome, text, timers and
buttons: square boxes are detected, grouped by size, and the group that forms
a tight lattice is taken as the matrix. A second group of equally sized boxes
arranged in one or two rows is taken as the answer options.
"""
from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np


@dataclass
class Box:
    x: int
    y: int
    w: int
    h: int

    @property
    def cx(self) -> float:
        return self.x + self.w / 2

    @property
    def cy(self) -> float:
        return self.y + self.h / 2

    @property
    def area(self) -> int:
        return self.w * self.h

    def contains(self, o: "Box", pad: int = 2) -> bool:
        return self.x - pad <= o.x and self.y - pad <= o.y and o.x + o.w <= self.x + self.w + pad and o.y + o.h <= self.y + self.h + pad

    def iou(self, o: "Box") -> float:
        ix = max(0, min(self.x + self.w, o.x + o.w) - max(self.x, o.x))
        iy = max(0, min(self.y + self.h, o.y + o.h) - max(self.y, o.y))
        inter = ix * iy
        return inter / float(self.area + o.area - inter)

    def as_list(self) -> list[int]:
        return [int(self.x), int(self.y), int(self.w), int(self.h)]


class DetectionError(Exception):
    def __init__(self, stage: str, message: str, region: Box | None = None, boxes: list[Box] | None = None):
        super().__init__(message)
        self.stage = stage
        self.region = region
        self.boxes = boxes or []


def square_boxes(binary: np.ndarray, edges: np.ndarray) -> list[Box]:
    """Candidate square boxes (cell borders / option borders)."""
    h, w = binary.shape
    min_side = max(20, int(0.03 * min(h, w)))
    max_side = int(0.6 * min(h, w))
    found: list[Box] = []
    closed = cv2.morphologyEx(edges, cv2.MORPH_CLOSE, np.ones((3, 3), np.uint8))
    for src in (binary, closed):
        contours, _ = cv2.findContours(src, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
        for c in contours:
            x, y, bw, bh = cv2.boundingRect(c)
            if bw < min_side or bh < min_side or bw > max_side or bh > max_side:
                continue
            if not 0.85 <= bw / bh <= 1.18:
                continue
            area = cv2.contourArea(c)
            if area < 0.85 * bw * bh:
                continue
            approx = cv2.approxPolyDP(c, 0.03 * cv2.arcLength(c, True), True)
            if len(approx) != 4:
                continue
            found.append(Box(x, y, bw, bh))
    # Deduplicate (outer/inner edge of the same border line): keep the inner box.
    found.sort(key=lambda b: b.area)
    out: list[Box] = []
    for b in found:
        if any(b.iou(o) > 0.75 or (o.contains(b, 4) and b.contains(o, 8)) for o in out):
            continue
        out.append(b)
    return out


def _cluster_1d(values: list[float], tol: float) -> list[float]:
    centers: list[list[float]] = []
    for v in sorted(values):
        if centers and abs(v - np.mean(centers[-1])) <= tol:
            centers[-1].append(v)
        else:
            centers.append([v])
    return [float(np.mean(c)) for c in centers]


def _size_groups(boxes: list[Box]) -> list[list[Box]]:
    groups: list[list[Box]] = []
    for b in sorted(boxes, key=lambda b: -b.w):
        for g in groups:
            ref = np.median([x.w for x in g])
            if abs(b.w - ref) <= 0.1 * ref and abs(b.h - np.median([x.h for x in g])) <= 0.1 * ref:
                g.append(b)
                break
        else:
            groups.append([b])
    return [g for g in groups if len(g) >= 2]


@dataclass
class Lattice:
    rows: int
    cols: int
    xs: list[float]
    ys: list[float]
    size: float
    cells: dict[tuple[int, int], Box]

    @property
    def missing(self) -> list[tuple[int, int]]:
        return [(r, c) for r in range(self.rows) for c in range(self.cols) if (r, c) not in self.cells]

    def slot_box(self, r: int, c: int) -> Box:
        s = int(round(self.size))
        return Box(int(round(self.xs[c] - s / 2)), int(round(self.ys[r] - s / 2)), s, s)

    @property
    def bbox(self) -> Box:
        s = self.size
        x0 = min(self.xs) - s / 2
        y0 = min(self.ys) - s / 2
        return Box(int(x0), int(y0), int(max(self.xs) - min(self.xs) + s), int(max(self.ys) - min(self.ys) + s))


def fit_lattice(group: list[Box]) -> Lattice | None:
    size = float(np.median([b.w for b in group]))
    tol = 0.25 * size
    xs = _cluster_1d([b.cx for b in group], tol)
    ys = _cluster_1d([b.cy for b in group], tol)
    # Fill gaps: a missing column/row appears as a spacing twice the pitch.
    def fill(centers: list[float]) -> list[float]:
        if len(centers) < 2:
            return centers
        pitch = min(np.diff(centers))
        out = [centers[0]]
        for v in centers[1:]:
            while v - out[-1] > 1.5 * pitch:
                out.append(out[-1] + pitch)
            out.append(v)
        return out
    xs, ys = fill(xs), fill(ys)
    cells: dict[tuple[int, int], Box] = {}
    for b in group:
        c = int(np.argmin([abs(b.cx - x) for x in xs]))
        r = int(np.argmin([abs(b.cy - y) for y in ys]))
        if abs(b.cx - xs[c]) <= tol and abs(b.cy - ys[r]) <= tol:
            cells[(r, c)] = b
    return Lattice(len(ys), len(xs), xs, ys, size, cells)


def pitch_ratio(lat: Lattice) -> float:
    pitches = []
    if len(lat.xs) > 1:
        pitches.append(float(np.median(np.diff(lat.xs))))
    if len(lat.ys) > 1:
        pitches.append(float(np.median(np.diff(lat.ys))))
    return (min(pitches) / lat.size) if pitches else 99.0


@dataclass
class Layout:
    matrix: Lattice
    options: list[Box]
    missing: tuple[int, int] | None
    candidates: list[Box]
    inferred_options: int = 0


def detect_layout(binary: np.ndarray, edges: np.ndarray) -> Layout:
    boxes = square_boxes(binary, edges)
    if len(boxes) < 4:
        raise DetectionError("matrix", "Too few box-like regions found; no matrix grid visible.", None, boxes)
    groups = _size_groups(boxes)
    best: Lattice | None = None
    best_group: list[Box] | None = None
    for g in groups:
        lat = fit_lattice(g)
        if lat is None or lat.rows * lat.cols < 4:
            continue
        # A matrix tiles tightly: pitch close to the cell size.
        if lat.rows >= 2 and lat.cols >= 2 and pitch_ratio(lat) <= 1.35:
            if len(lat.cells) < lat.rows * lat.cols - 2:
                continue
            if best is None or lat.size * len(lat.cells) > best.size * len(best.cells):
                best, best_group = lat, g
    if best is None:
        region = max(groups, key=lambda g: sum(b.area for b in g)) if groups else boxes
        xs = [b.x for b in region] + [b.x + b.w for b in region]
        ys = [b.y for b in region] + [b.y + b.h for b in region]
        raise DetectionError(
            "matrix",
            "No tightly packed grid of equally sized cells was found.",
            Box(min(xs), min(ys), max(xs) - min(xs), max(ys) - min(ys)),
            boxes,
        )
    matrix = best
    mb = matrix.bbox
    # Answer options: another group of equal boxes, outside the matrix, in 1-2 rows.
    opt_best: list[Box] | None = None
    for g in groups:
        if g is best_group:
            continue
        g = [b for b in g if not mb.contains(b, 6)]
        if not 3 <= len(g) <= 8:
            continue
        # Must not be nested inside other candidate boxes (objects inside options)
        if any(any(o.contains(b, 2) and o.area > 1.5 * b.area for o in boxes) for b in g):
            continue
        rows = _cluster_1d([b.cy for b in g], 0.3 * np.median([b.h for b in g]))
        if len(rows) > 2:
            continue
        if opt_best is None or np.median([b.w for b in g]) > np.median([b.w for b in opt_best]):
            opt_best = g
    if opt_best is None:
        raise DetectionError("options", "The matrix was found but no row of answer options was detected.", mb, boxes)
    # Order options: top-to-bottom rows, then left-to-right
    rows = _cluster_1d([b.cy for b in opt_best], 0.3 * np.median([b.h for b in opt_best]))
    options = sorted(opt_best, key=lambda b: (int(np.argmin([abs(b.cy - r) for r in rows])), b.cx))
    options, inferred = _fill_option_gaps(options, rows)
    missing = matrix.missing
    return Layout(matrix, options, missing[0] if len(missing) == 1 else None, boxes, inferred)


def _fill_option_gaps(options: list[Box], rows: list[float]) -> tuple[list[Box], int]:
    """An option whose border is broken (e.g. by a large figure) leaves a gap in
    the evenly spaced row. Insert an inferred box of the same size there."""
    if len(options) < 3:
        return options, 0
    out: list[Box] = []
    inferred = 0
    for r in rows:
        row = [b for b in options if int(np.argmin([abs(b.cy - x) for x in rows])) == rows.index(r)]
        row.sort(key=lambda b: b.cx)
        if len(row) < 2:
            out.extend(row)
            continue
        pitch = float(np.min(np.diff([b.cx for b in row])))
        filled = [row[0]]
        for b in row[1:]:
            while b.cx - filled[-1].cx > 1.5 * pitch:
                p = filled[-1]
                filled.append(Box(int(round(p.x + pitch)), p.y, p.w, p.h))
                inferred += 1
            filled.append(b)
        out.extend(filled)
    return out, inferred
